### PR TITLE
Perplexity gpu mem optimization

### PR DESCRIPTION
**Pull request by mosheraboh**

### CODE REVIEW

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py [hunk 1]
```diff
@@ -96,6 +103,10 @@ def _perplexity_update(
         mask = torch.ones_like(target, dtype=torch.bool)
 
     preds = preds[:, target].diagonal()[mask]
+    # avoid from overflow
+    if preds.dtype == torch.float16:
+        preds = preds.to(torch.float32)
+        preds = torch.clamp(preds, min=1e-10)
```
[the remark this answers is not in the document]
Comment (SagiPolaczek) [reply fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:109]: Plus, I think that values in `(0, 1e-10)` cannot be exist in `float16`, so why clamping that after moving to `float32` ?

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py [hunk 2]
```diff
@@ -106,6 +117,10 @@ def _perplexity_compute(
     log_probs: List[np.ndarray],
     token_num: List[np.ndarray],
 ) -> float:
+    # avoid from overflow on large epochs
+    log_probs = [e.astype(np.float64) for e in log_probs]
```
Comment (SagiPolaczek) [inline fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:121]: numerical stability?

File: fuse/eval/metrics/metrics_common.py [hunk 3]
```diff
@@ -239,15 +243,41 @@ def sync_ids(ids: List[Tuple[str, int]]) -> List[Any]:
         return data_list
 
     @staticmethod
-    def _df_dict_apply(data: pd.Series, func: Callable) -> pd.Series:
-        result = func(NDict(data.to_dict()))
-        return pd.Series(result.flatten())
+    def _df_dict_apply(
+        data: pd.Series, func: Callable, batch: bool = False
+    ) -> pd.Series:
+        if batch:
```
Comment (SagiPolaczek) [inline fuse/eval/metrics/metrics_common.py:249]: Not sure I understood the use-case: `data` is always a single data point (e.g. series) and `func` might be perform on a batch level? And where `func` expects a batch you pass `batch=True`?

I'm suggesting to rename `batch` or add a short description to make it clearer :))

**Review by SagiPolaczek — APPROVED** [review]

**Review by SagiPolaczek — APPROVED** [review]
LGTM!
I added some questions inline :)

Could you please also describe the GPU mem opt part? Maybe how did you find it and what exactly fixing it (in high-level of course).

File: fuse/eval/examples/examples_seq_gen.py [hunk 4]
```diff
@@ -47,11 +46,7 @@ def example_seq_gen_0(seed: int = 1234) -> Dict[str, Any]:
         [
             (
                 "perplexity",
-                CI(
```
[the remark this answers is not in the document]
Comment (mosheraboh) [reply fuse/eval/examples/examples_seq_gen.py:50]: We do cover CI in other examples I hope.
I removed it because the perplexity now is computed at the batch level (more optimized) and I'm not sure it's a metric we would want a CI for.
We can support it back when necessary.

File: fuse/eval/metrics/metrics_common.py (hunk 3 above)
Comment (mosheraboh) [reply fuse/eval/metrics/metrics_common.py:249]: The metric has some preprocessing func in a batch level.
Here we are not running a training loop,, to avoid duplication, we run the function on each sample as if it was a batch. To do that we are converting single sample to batch, call the function and then squeezed it back to sample.
As I mentioned, those are dark places in fuse, that I would refactor If we had some time to spend on it.

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py [hunk 5]
```diff
@@ -34,22 +34,27 @@ def __init__(
         **kwargs: dict,
     ) -> None:
         super().__init__(
-            preds=preds,
-            target=target,
-            metric_per_batch_func=partial(
-                _perplexity_update, ignore_index=ignore_index
+            log_probs="log_probs",  # collect log_probs - output of  _perplexity_update
```
[the remark this answers is not in the document]
Comment (mosheraboh) [reply fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:37]: I agree with you.
But it's indeed out of PR scope.

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mosheraboh) [reply fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:109]: I'm getting absolute zeros in float16 precision which causes log() to be infinity.

File: fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py (hunk 2 above)
Comment (mosheraboh) [reply fuse/eval/metrics/sequence_gen/metrics_seq_gen_common.py:121]: It might be large numbers with many GPUS so I took the safe side here.
